hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/NativeSparkDataSet.java

```diff
@@ -1114,64 +1116,98 @@ public static boolean joinProjectsOnlyLeftTableColumns(JoinType joinType) {
-        ValueRow valueRow=new ValueRow(1);
-        valueRow.setColumn(1,new SQLLongint(context.getRecordsWritten()));
-        return new SparkDataSet<>(SpliceSpark.getContext().parallelize(Collections.singletonList(valueRow), 1));
+        return getRowsWritten(context);
     }
 
     @SuppressWarnings({ "unchecked", "rawtypes" })
-    public DataSet<ExecRow> writeAvroFile(DataSetProcessor dsp,
-                                          int[] partitionBy,
-                                          String location,
-                                          String compression,
-                                          OperationContext context) throws StandardException
-    {
-        compression = SparkDataSet.getAvroCompression(compression);
-        DataFrameWriter writer = getDataFrameWriter(partitionBy, compression, context);
-        writer.mode(SaveMode.Append).format("com.databricks.spark.avro").save(location);
-        ValueRow valueRow=new ValueRow(1);
-        valueRow.setColumn(1,new SQLLongint(context.getRecordsWritten()));
-        return new SparkDataSet<>(SpliceSpark.getContext().parallelize(Collections.singletonList(valueRow), 1));
+    public DataSet<ExecRow> writeTextFile(int[] partitionBy, String location, CsvOptions csvOptions,
+                                          OperationContext context) throws StandardException {
+        try( CountingListener counter = new CountingListener(context) ) {
+            getDataFrameWriter(partitionBy, context)
+                    .options(getCsvOptions(csvOptions))
+                    .csv(location);
+        }
+        return getRowsWritten(context);
     }
-
-
     @SuppressWarnings({ "unchecked", "rawtypes" })
     public DataSet<ExecRow> writeORCFile(int[] baseColumnMap, int[] partitionBy, String location,  String compression,
                                                     OperationContext context) throws StandardException {
-        DataFrameWriter writer = getDataFrameWriter(partitionBy, compression, context);
-        writer.mode(SaveMode.Append).orc(location);
-        ValueRow valueRow=new ValueRow(1);
-        valueRow.setColumn(1,new SQLLongint(context.getRecordsWritten()));
-        return new SparkDataSet<>(SpliceSpark.getContext().parallelize(Collections.singletonList(valueRow), 1));
+        try( CountingListener counter = new CountingListener(context) ) {
+            getDataFrameWriter(partitionBy, context)
+                    .option(SPARK_COMPRESSION_OPTION, compression)
+                    .orc(location);
+        }
+        return getRowsWritten(context);
     }
 
-    private DataFrameWriter getDataFrameWriter(int[] partitionBy, String compression, OperationContext context) throws StandardException {
+    class CountingListener extends SparkListener implements AutoCloseable
+    {
+        OperationContext context;
+        SparkContext sc;
+        String uuid;
+        List<Integer> stageIdsToWatch;
```

Comment:
Can you make it a Set? Since we are going to call lots of `.contains()` on it I think that makes sense.

Reply:
yes will do